File: src/main/scala/defo/ls/CumulativeTree.scala
package srte.ls

class CumulativeTree(size: Int) {
  
  private[this] val weights = new Array[Double](size*2)
  
  def totalWeight: Double = weights(1)
  
  def changeWeight(value: Int, weight: Double): Unit = { 
    var i = size + value
    weights(i) = weight
    i = i >> 1
    while(i > 0) {     
      val l = i << 1
      val r = l + 1     
      weights(i) = weights(l) + weights(r) 
      i = i >> 1
    }
  }
  
  def get(weight: Double): Int = {
    var i = 1
    var w = weight
    while (i < size) {
      val l = i << 1
      val r = l + 1 
      if (w < weights(l)) i = l
      else {
        w = w - weights(l)
        i = r
      }
    }
    i - size
  }
}

object Test extends App {
  val t = new CumulativeTree(4)
  t.changeWeight(0, 2)
  t.changeWeight(1, 1)
  t.changeWeight(2, 4)
  t.changeWeight(3, 8)
  println(t.totalWeight)
  for (i <- 0 to t.totalWeight.toInt) {
    println(t.get(i))
  }
}

File: src/main/java/defo/util/CumulativeTree.java
package defo.util;

public class CumulativeTree {

  private final int size;
  private double[] weights;
  
  public CumulativeTree(int size) {
    this.size = size;
    this.weights = new double[size * 2];
  }
  
  public double totalWeight() {
    return weights[1];
  }
  
  public void setWeight(int elemId, double weight) {
    int i = size + elemId;
    weights[i] = weight;
    i = i >> 1;
    while (i > 0) {
      final int left = i << 1;
      final int right = left + 1;
      weights[i] = weights[left] + weights[right];
      i = i >> 1;
    }
  }
  
  public int get(double weight) {
    int i = 1;
    double w = weight;
    while (i < size) {
      final int left = i << 1;
      final int right = left + 1;
      if (w < weights[left]) {
        i = left;
      } else {
        w -= weights[left];
        i = right;
      }
    }
    return i - size;
  }
}
